socorro/external/postgresql/new_crash_source.py

```diff
@@ -0,0 +1,242 @@
+
+    #--------------------------------------------------------------------------
+    def get_raw_dump(self, crash_id, name=None):
+        return self._implementation.get_raw_dump(crash_id, name)
+
+    #--------------------------------------------------------------------------
+    def get_raw_dumps(self, crash_id):
+        return self._implementation.get_raw_dumps(crash_id)
+
+    #--------------------------------------------------------------------------
+    def get_raw_dumps_as_files(self, crash_id):
+        return self._implementation.get_raw_dumps_as_files(crash_id)
+
+    #--------------------------------------------------------------------------
+    def get_processed(self, crash_id):
+        return self._implementation.get_processed(crash_id)
+
+    #--------------------------------------------------------------------------
+    def get_unredacted_processed(self, crash_id):
+        return self._implementation.get_unredacted_processed(crash_id)
+
+    #--------------------------------------------------------------------------
+    def remove(self, crash_id):
+        self._implementation.remove(crash_id)
+
+
+#==============================================================================
+class PGPVNewCrashSource(PGQueryNewCrashSource):
+    required_config = Namespace()
+    required_config.crash_id_query = change_default(
+        PGQueryNewCrashSource,
+        'crash_id_query',
+        "select uuid from reports "
+        "where ( "
+        "date_processed >= %s and date_processed < %s "
+        "and ({pv_where_clause}))"
+    )
+    required_config.add_option(
+        "product_version_list",
+        doc='list of the form "p1, v1a, v1b; p2, v2a, v2b; ..."',
+        default='Firefox',
+        from_string_converter=two_level_nested_list_converter(),
+        likely_to_be_changed=True
+    )
+    required_config.add_option(
+        'date',
+        doc="a date in the form YYYY-MM-DD",
+        default=(utc_now() - timedelta(1)).date(),
+        from_string_converter=string_to_datetime
+    )
+
+    #--------------------------------------------------------------------------
+    def __init__(self, config, name, quit_check_callback=None):
+        super(PGPVNewCrashSource, self).__init__(
+            config,
+            name,
+            quit_check_callback
+        )
+        where_clause_element = "(product = %s and version = ANY(%s))"
+        pv_where_clause = (
+            ' or '.join(
+                [where_clause_element] * len(config.product_version_list)
+            )
+        )
+        self.crash_id_query = config.crash_id_query.format(
+            pv_where_clause=pv_where_clause,
+        )
+        self.data = [
+            config.date,
+            config.date + timedelta(1)
+        ]
+
+        def flatten(list_of_tuples):
+            for x in iter(list_of_tuples):
+                if isinstance(x, Sequence) and not isinstance(x, basestring):
```

Comment:
I'm iterating over a sequence of sequences [ 'Firefox', ('v1', 'v2', 'v3')] with the aim of flattening it to ['Firefox', 'v1', 'v2', 'v3'].  Since a string is also a sequence, if I did not make the special case, it would flatten to [ 'F', 'i', 'r', 'e', 'f', 'o', 'x', 'v1', 'v2', v3' ]